parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -0,0 +1,169 @@
+  a <> b =
+    DefinitionDiffs
+      { termDiffs = termDiffs a <> termDiffs b,
+        typeDiffs = typeDiffs a <> typeDiffs b
+      }
+
+instance Monoid DefinitionDiffs where
+  mempty = DefinitionDiffs mempty mempty
+
+-- | A tree of local diffs. Each node of the tree contains the definition diffs at that path.
+newtype TreeDiff = TreeDiff
+  { unTreeDiff :: Cofree (Map NameSegment) DefinitionDiffs
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance Semigroup TreeDiff where
+  TreeDiff (a :< as) <> TreeDiff (b :< bs) =
+    TreeDiff $ (a <> b) :< (Map.unionWith mergeCofrees as bs)
+    where
+      mergeCofrees x y = unTreeDiff (TreeDiff x <> TreeDiff y)
+
+instance Monoid TreeDiff where
+  mempty = TreeDiff (mempty :< mempty)
+
+instance Lens.AsEmpty TreeDiff where
+  _Empty = Lens.only mempty
+
+-- | A summary of a 'TreeDiff', containing all names added and removed.
+-- Note that there isn't a clear notion of a name "changing" since conflicts might muddy the notion
+-- by having multiple copies of both the from and to names, so we just talk about adds and
+-- removals instead.
+data NameChanges = NameChanges
+  { termNameAdds :: [(Name, Referent)],
+    termNameRemovals :: [(Name, Referent)],
+    typeNameAdds :: [(Name, Reference)],
+    typeNameRemovals :: [(Name, Reference)]
+  }
+
+instance Semigroup NameChanges where
+  (NameChanges a b c d) <> (NameChanges a2 b2 c2 d2) =
+    NameChanges (a <> a2) (b <> b2) (c <> c2) (d <> d2)
+
+instance Monoid NameChanges where
+  mempty = NameChanges mempty mempty mempty mempty
+
+-- | Diff two Branches, returning a tree containing all of the changes
+diffBranches :: forall m. Monad m => Branch m -> Branch m -> m TreeDiff
+diffBranches from to = do
+  let termDiffs = diffMap (terms from) (terms to)
+  let typeDiffs = diffMap (types from) (types to)
+  let defDiff = DefinitionDiffs {termDiffs, typeDiffs}
+  rec <- do
+    Align.align (children from) (children to)
+      & wither
+        ( \case
+            This ca -> do
+              -- TODO: For the names index we really don't need to know which exact
+              -- names were removed, we just need to delete from the index using a
+              -- prefix query, this would be faster than crawling to get all the deletes.
+              removedChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
+            That ca -> do
+              newChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                -- This child didn't change.
+                pure Nothing
+              | otherwise -> do
+                fromChildBranch <- Causal.value fromC
+                toChildBranch <- Causal.value toC
+                diffBranches fromChildBranch toChildBranch >>= \case
+                  Lens.Empty -> pure Nothing
```

Comment:
Is this case possible?

Reply:
At the moment, yes, but I think only because we don't check metadata or patches yet, so it's possible the hash is different because of metadata but the currently computed diff is empty.